lib/Backends/Interpreter/InterpreterDeviceMemoryHelper.cpp

```diff
@@ -0,0 +1,85 @@
+/**
+ * Copyright (c) 2017-present, Facebook, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#include "InterpreterDeviceMemoryHelper.h"
+
+using namespace glow;
+
+InterpreterDeviceMemoryHelper::InterpreterDeviceMemoryHelper() {}
+
+InterpreterDeviceMemoryHelper::~InterpreterDeviceMemoryHelper() {
+  for (const auto &pair : tensors_) {
+    delete pair.second;
+  }
+  tensors_.clear();
+}
+
+void InterpreterDeviceMemoryHelper::allocateTensor(channel_id_type channelId,
+                                                   TypeRef tensorType) {
+  std::lock_guard<std::mutex> tGuard(tensorsGuard_);
```

Comment:
If anywhere else the order of lock will be reversed it will cause the dead-lock.
Better to lock only for the particular resource and release lock right away.

`std::unique_lock<std::mutex> tGuard(tensorsGuard_);

  assert(tensors_.find(channelId) == tensors_.end() &&
         "Tensor already allocated for channel ID");

  tensors_.insert(std::make_pair(channelId, new Tensor(tensorType)));
 
 tGuard.unlock();

 std::unique_lock<std::mutex> sGuard(transferStatusGuard_);

 transferStatus_.insert(std::make_pair(channelId, false));
 `